### DISCUSSION TITLE
Memory Leak in forge-pull?

### DISCUSSION THREAD

**Discussion by t-zuehlsdorff** [magit/forge#698] [Support]
Aloha,

thank you for your release! I was finally able to use forge successfully in several small projects and it is great! :) 

However, i found a bug when trying to use a mid-size project. Doing a `forge-pull` in small projects with ~ 100 issues and MRs is fine. But when i do this for ~ 1.200 issues and MRs emacs crashes. I can see that the issues got fetched and the MRs are getting pulled in. At some point in time my Laptop become slower, later emacs will not render anymore and get killed.

I saw it eating up to 32 GB of RAM before the OOM-Killer did his job.

I am using forge 0.4.1 on FreeBSD and emacs 24.2. I am pulling from a self hosted GitLab.

I am not sure how i can provide helpful information to debug this issue. If you can guide me i will provide everything you need :)

Best,
Torsten

**Comment by tarsius** [comment]
> I am using forge 0.4.1 on FreeBSD and emacs 24.2. I am pulling from a self hosted GitLab.

I'm afraid that is a *wontfix*.

Forge 0.4 requires Emacs 27.1. It might kinda work with older releases, but the use of an ancient release could very well be the cause of this issue. One result of this is that the new *builtin* and *module* back-ends of Emacsql cannot be used. Instead the legacy *custom external process* back-end has to be used. That back-end will be removed soon, possibly as early as Monday.

Additionally support for Gitlab is limited. Much fewer Forge users use it with Gitlab than with Github, so I cannot justify putting as much effort into supporting the former.

I recommend you update to Emacs 29.4. That *might* just by-pass this issue altogether (and has many other benefits).

**Reply by t-zuehlsdorff** [reply]
@tarsius I am sorry, i mistyped the version number. I am using Emacs 29.4.

**Reply by t-zuehlsdorff** [reply]
@tarsius I played a little bit around. The issue seems to be in the pulling of merge requests. The memory usage of Emacs goes from 150 MB to 3960 MB when pulling 135 Requests.

**Reply by tarsius** [reply]
> I am using Emacs 29.4.

Even when using Emacs 29.4 you could end up not using one of the new back-ends. Check with `(class-of (oref (forge-db) connection))`.

> The memory usage of Emacs goes from 150 MB to 3960 MB when pulling 135 Requests.

I fetched a repository with 1100 issues and 1700 pull-requests from Gitlab. That temporarily increased `emacs`'s memory usage from 0.8% to 1.1% (of 32GB).

Have you tried with other repositories from your Gitlab instance and gitlab.com?

**Reply by pkryger** [reply]
@t-zuehlsdorff this is a long-shot but what are values of `gc-cons-threshold` and `gc-cons-percentage`?

**Reply by t-zuehlsdorff** [reply]
@pkryger Values are:
* `gc-cons-threshold`: 800000
* `gc-cons-percentage`: 0.1

@tarsius 
> Even when using Emacs 29.4 you could end up not using one of the new back-ends. Check with (class-of (oref (forge-db) connection)).

For this i get `emacsql-sqlite-builtin-connection`.

**Comment by tarsius** [comment]
Off-topic: @t-zuehlsdorff My emails to you no longer go through for some reason.

**Reply by t-zuehlsdorff** [reply]
@tarsius I broke the mailserver without noticing around 2024-12-23. It is fixed right now. :) Sorry!
